### PR TITLE
deprecate sprandbool + add new syntax for creating random sparse matrix

### LINKED ISSUE #11688: Deprecate sprandbool
Just like we deprecated `randbool`, we should probably deprecate `sprandbool`. While doing so, it would be nice to extend the `sprand` API to generate sparse matrices with all the element types the RNG supports, similar to the dense `rand`.

### PR DESCRIPTION
**Pull request by KristofferC**
This mimics the one already existing for `rand`.

Now:

``` jl
julia> rand(Float32, 3, 3)
3×3 Array{Float32,2}:
 0.938487  0.200827  0.746491
 0.566157  0.650916  0.466414
 0.408861  0.166284  0.648508

julia> sparsity = 0.5
0.5

julia> sprand(Float32, 3, 3, sparsity)
3×3 sparse matrix with 4 Float32 nonzero entries:
    [2, 1]  =  0.181532
    [3, 2]  =  0.51335
    [1, 3]  =  0.630208
    [2, 3]  =  0.632587
```

[edit by tkelman: closes #11688]

### CODE REVIEW

**Comment by KristofferC** [comment]
Probably makes sense to deprecate sprandbool here too since now it is just:

```jl
julia> sprand(Bool, 3, 3, 0.5)
3×3 sparse matrix with 5 Bool nonzero entries:
	[2, 1]  =  false
	[1, 2]  =  false
	[2, 2]  =  false
	[3, 2]  =  true
	[3, 3]  =  true
```

**Comment by KristofferC** [comment]
I'm a bit confused. Is it intended that `sprandbool` currently only producing trues?

```jl
julia> sprandbool(3,3, 0.5)
3x3 sparse matrix with 5 Bool entries:
	[1, 1]  =  true
	[3, 1]  =  true
	[1, 2]  =  true
	[3, 2]  =  true
	[3, 3]  =  true
```

I would have thought it should behave like:

```jl
julia> rand(Bool, 3,3)
3x3 Array{Bool,2}:
 false  false   true
  true  false   true
 false   true  false
```

**Comment by KristofferC** [comment]
Actually the above is wrong since `false` shouldn't be stored.

File: base/sparse/sparsevector.jl [hunk 1]
```diff
@@ -356,17 +356,17 @@ function sprand(r::AbstractRNG, n::Integer, p::AbstractFloat, rfn::Function)
     SparseVector(n, I, V)
 end
 
+sprand{T}(::Type{T}, n::Integer, p::AbstractFloat) = sprand(GLOBAL_RNG, n, p, rand, T)
 sprand{T}(n::Integer, p::AbstractFloat, ::Type{T}) = sprand(GLOBAL_RNG, n, p, rand, T)
```
Comment (KristofferC) [inline base/sparse/sparsevector.jl]: Should this one be deprecated as well? It doesn't correspond to the dense case.

File: base/deprecated.jl [hunk 2]
```diff
@@ -1078,6 +1078,18 @@ end
...
 
+"""
+    sprandbool(m[,n],p)
+Create a random `m` by `n` sparse boolean matrix or length `m` sparse boolean
+vector with the specified (independent) probability `p` of any entry being
+`true`.
+"""
+function sprandbool end
+@deprecate sprandbool(m::Integer, n::Integer, density::AbstractFloat) sprand(Bool, m, n, density)
+@deprecate sprandbool(r::AbstractRNG, m::Integer, n::Integer, density::AbstractFloat) sprand(r, m, n, density, (r, i) -> ones(Bool, i))
```
Comment (KristofferC) [inline base/deprecated.jl]: This is a bit ugly but I don't think it is used much.

**Comment by tkelman** [comment]
Needs a doc update for the new `sprand` syntax with eltype first. It's a little annoying that eltype comes last in the existing version of `sprand` if you want to provide both an RNG and an eltype. Ideally we could make the whole thing more consistent with the way `rand` works when you want to provide just one or both (but with the additional density parameter).

**Comment by KristofferC** [comment]
Yes, I will make stuff more consistent between `rand` and `sprand` but then there will have to be some deprecations for `sprand` too

**Comment by KristofferC** [comment]
Updated. Some usage examples:

```jl
julia> sprand(Bool, 3, 3, 0.3)
3×3 sparse matrix with 2 Bool nonzero entries:
	[3, 1]  =  true
	[1, 3]  =  true

julia> sprand(Base.GLOBAL_RNG, Bool, 3, 3, 0.3)
3×3 sparse matrix with 4 Bool nonzero entries:
	[1, 1]  =  true
	[3, 1]  =  true
	[1, 3]  =  true
	[2, 3]  =  true

julia> sprand(Base.GLOBAL_RNG, Float32, 3, 3, 0.3)
3×3 sparse matrix with 2 Float32 nonzero entries:
	[3, 2]  =  0.954499
	[2, 3]  =  0.00251913

julia> sprand(Float64, 3, 3, 0.3)
3×3 sparse matrix with 3 Float64 nonzero entries:
	[2, 1]  =  0.960525
	[3, 1]  =  0.990573
	[1, 3]  =  0.475681

julia> sprand(Float64, 2, 0.3)
Sparse vector of length 2 with 1 Float64 nonzero entries:
  [1]  =  0.446653
```

File: base/sparse/sparsematrix.jl [hunk 3]
```diff
@@ -894,12 +894,12 @@ function sprand_IJ(r::AbstractRNG, m::Integer, n::Integer, density::AbstractFloa
...
 
 """
-    sprand([rng],m,[n],p::AbstractFloat,[rfn])
+    sprand([rng],[type],m,[n],p::AbstractFloat,[rfn])
 
 Create a random length `m` sparse vector or `m` by `n` sparse matrix, in
 which the probability of any element being nonzero is independently given by
 `p` (and hence the mean density of nonzeros is also exactly `p`). Nonzero
-values are sampled from the distribution specified by `rfn`. The uniform
+values are sampled from the distribution specified by `rfn` and have the type `type`. The uniform
```
Comment (tkelman) [inline base/sparse/sparsematrix.jl]: run genstdlib

looks really good, thanks for getting to this before I did

Comment (KristofferC) [inline base/sparse/sparsematrix.jl]: wow, you are fast.. just ran it.

**Comment by KristofferC** [comment]
What is the policy regarding documentation? Should documentation for deprecated methods be removed?

**Comment by tkelman** [comment]
I think that's what we've been doing. The deprecation warning will point to the new syntax and the new syntax should be as well or better documented than the old syntax.

File: base/sparse/sparsematrix.jl (hunk 3 above)
Comment (KristofferC) [inline base/sparse/sparsematrix.jl]: think I need to rerun make for the actual docstring to get updated though..

Comment (tkelman) [inline base/sparse/sparsematrix.jl]: yeah `make docs` is the more precise way that will also rerun bootstrap to get the updated docstring into the system image

**Comment by KristofferC** [comment]
I built up a bit of unnecessary CI queue which could possibly be killed.

**Comment by nalimilan** [comment]
I've cancelled the duplicate builds.

**Comment by tkelman** [comment]
there are hooks that fail fast on redundant builds automatically, so not a big deal. manually cancelling saves maybe a minute or two tops.

**Comment by nalimilan** [comment]
@tkelman The issue is that the obsolete builds had already started, and AFAIK they aren't cancelled automatically in that case.

**Comment by tkelman** [comment]
ah yes good call

**Comment by KristofferC** [comment]
Error is OSX timeout.

**Comment by tkelman** [comment]
Last few places it should be cleaned up, from `git grep`:
```
base/exports.jl:1462:    sprandbool,
base/sparse.jl:34:    sprand, sprandbool, sprandn, spzeros, symperm, nnz
contrib/BBEditTextWrangler-julia.plist:1073:            <string>sprandbool</string>
```
then we're good to merge

**Comment by KristofferC** [comment]
Doesn't it still need to be exported? Or does the deprecated macro export?

**Comment by tkelman** [comment]
the deprecate macro should be handling the exporting, I believe

**Comment by nalimilan** [comment]
> the deprecate macro should be handling the exporting, I believe

That's right (as documented in the comment at the top of deprecated.jl).

**Comment by KristofferC** [comment]
I accidentally amended the latest changes to the "remove documentation" commit. However, if things are fine now then it will be squashed by the one who merges anyway so it shouldn't matter much.

**Comment by KristofferC** [comment]
Anyway this is ready from my pov.

File: test/sparsedir/sparsevector.jl [hunk 4]
```diff
@@ -152,7 +152,7 @@ end
 let r1 = MersenneTwister(), r2 = MersenneTwister()
     @test sprand(r1, 100, .9) == sprand(r2, 100, .9)
     @test sprandn(r1, 100, .9) == sprandn(r2, 100, .9)
-    @test sprandbool(r1, 100, .9) == sprandbool(r2, 100, .9)
+    @test sprand(r1, Bool, 100, .9, ) == sprand(r2,  Bool, 100, .9)
```
Comment (tkelman) [inline test/sparsedir/sparsevector.jl]: any recollection why there's an extra comma here?

Comment (KristofferC) [inline test/sparsedir/sparsevector.jl]: Accident probably.

**Pull request merged**